actors/builtin/miner/deadlines.go

```diff
@@ -209,18 +223,45 @@ func AssignNewSectors(deadlines *Deadlines, partitionSize uint64, newSectors []u
 		}
 	}
 
-	// While there remain new sectors to assign, fill a new partition at each deadline in round-robin fashion.
-	// TODO WPOST (follow-up): fill less-full deadlines first, randomize when equally full.
-	targetDeadline := uint64(1)
+	// While there remain new sectors to assign, fill a new partition in one of the deadlines that is least full.
+	// Do this by maintaining a slice of deadline indexes sorted by partition count.
+	// Shuffling this slice to re-sort as weights change is O(n^2).
+	// For a large number of partitions, a heap would be the way to do this in O(n*log n), but for small numbers
+	// is probably overkill.
+	// A miner onboarding a monumental 1EiB of 32GiB sectors uniformly throughout a year will fill 40 partitions
+	// per proving period (40^2=1600). With 64GiB sectors, half that (20^2=400).
+	// TODO: randomize assignment among equally-full deadlines https://github.com/filecoin-project/specs-actors/issues/432
+
+	dlIdxs := make([]uint64, WPoStPeriodDeadlines)
+	for i := range dlIdxs {
+		dlIdxs[i] = uint64(i)
+	}
+
+	sortDeadlines := func() {
+		// Order deadline indexes by corresponding partition count (then secondarily by index) to form a queue.
+		sort.Slice(dlIdxs, func(i, j int) bool {
```

Comment:
I'd feel more comfortable if this was
```suggestion
		sort.SliceStable(dlIdxs, func(i, j int) bool {
```

Reply:
I did consider that, but note that ordering secondarily by deadline index makes this a total ordering, so stability is irrelevant.